include/dali/core/exec/tasking/task.h

```diff
@@ -0,0 +1,522 @@
+   */
+  template <typename F>
+  static SharedTask Create(F &&function, double priority = 0) {
+    return std::make_shared<Task>(std::forward<F>(function), priority);
+  }
+
+  /** Creates a task with a multiple results
+   *
+   * @param num_results the number of results produced by the task function;
+   *                    the special value `ScalarResult` changes the interpretation of the result.
+   * @param function    the callable object that defines the task;
+   *                    if num_result != ScalarValue, then the function can return
+   *                    - an iterable type (one on which std::begin and std::end can be called)
+   *                    - a tuple
+   *                    if num_result == ScalarValue, the function can return anything
+   * @param priority    the priority with which the task will be popped by the scheduler, once ready
+   */
+  template <typename F>
+  static SharedTask Create(int num_results, F &&function, double priority = 0) {
+    return std::make_shared<Task>(num_results, std::forward<F>(function), priority);
+  }
+
+  ~Task() {
+    assert(prev == nullptr);
+    assert(next == nullptr);
+    assert(state != TaskState::Running && state != TaskState::Destroyed);
+    state = TaskState::Destroyed;
+  }
+
+  TaskState state = TaskState::New;
+
+  /** The priorirt of the task; the higher, ths sooner a task is picked */
+  double Priority() const {
+    return priority_;
+  }
+
+  /** If true, the task can be immediately moved to execution */
+  bool Ready() {
+    return preconditions_.empty();
+  }
+
+  /** Adds a precondition
+   *
+   * Calling Succeed adds the waitable `w` to the task's list of preconditions.
+   * Duplicates are detected and ignored.
+   *
+   * The Waitable can be a Task that's already submitted, running or even complete.
+   */
+  Task *Succeed(const std::shared_ptr<Waitable> &w) {
+    if (state != TaskState::New)
+      throw std::logic_error(
+          "Cannot add a new dependency to a task that has been submitted for execution.\n");
+    if (std::find(preconditions_.begin(), preconditions_.end(), w) == preconditions_.end())
+      preconditions_.push_back(w);
+    return this;
+  }
+
+  /** Subscribes to other task's output value
+   *
+   * This function does two things:
+   * - adds a dependency (Succeed) on the producer task
+   * - creates a shared pointer to the producer task's output
+   *
+   * The producer must not have been submitted to the scheruler when Consume is called.
+   */
+  Task *Consume(const SharedTask &producer, int output_index = 0) {
+    if (producer->state != TaskState::New)
+      throw std::logic_error(
+          "Cannot subscribe to a result of a task that's been already submitted for execution.\n"
+          "If only ordering is required, use Succeed instead.");
+    inputs_.push_back(producer->results_.GetChecked(output_index));
+    Succeed(producer);
```

Comment:
```suggestion
    Succeed(producer);
    inputs_.push_back(producer->results_.GetChecked(output_index));
```
So in case of exception from `Succeed` we don't get inconsistent state.

Reply:
Done.